pkg/scheduler/framework/v1alpha1/interface.go

```diff
@@ -495,3 +496,84 @@ type FrameworkHandle interface {
+	Filter(pod *v1.Pod, nodes []*v1.Node) (filteredNodes []*v1.Node, failedNodesMap extenderv1.FailedNodesMap, err error)
+
+	// Prioritize based on extender-implemented priority functions. The returned scores & weight
+	// are used to compute the weighted score for an extender. The weighted scores are added to
+	// the scores computed by Kubernetes scheduler. The total scores are used to do the host selection.
+	Prioritize(pod *v1.Pod, nodes []*v1.Node) (hostPriorities *extenderv1.HostPriorityList, weight int64, err error)
+
+	// Bind delegates the action of binding a pod to a node to the extender.
+	Bind(binding *v1.Binding) error
+
+	// IsBinder returns whether this extender is configured for the Bind method.
+	IsBinder() bool
+
+	// IsInterested returns true if at least one extended resource requested by
+	// this pod is managed by this extender.
+	IsInterested(pod *v1.Pod) bool
+
+	// ProcessPreemption returns nodes with their victim pods processed by extender based on
+	// given:
+	//   1. Pod to schedule
+	//   2. Candidate nodes and victim pods (nodeToVictims) generated by previous scheduling process.
+	// The possible changes made by extender may include:
+	//   1. Subset of given candidate nodes after preemption phase of extender.
+	//   2. A different set of victim pod for every given candidate node after preemption phase of extender.
+	ProcessPreemption(
+		pod *v1.Pod,
+		nodeToVictims map[string]*extenderv1.Victims,
+		nodeInfos NodeInfoLister,
+	) (map[string]*extenderv1.Victims, error)
+
+	// SupportsPreemption returns if the scheduler extender support preemption or not.
+	SupportsPreemption() bool
+
+	// IsIgnorable returns true indicates scheduling should not fail when this extender
+	// is unavailable. This gives scheduler ability to fail fast and tolerate non-critical extenders as well.
+	IsIgnorable() bool
+}
+
+// PreemptHandle incorporates all needed logic to run preemption logic.
+// It will be passed as an arg to PostFilter().
+type PreemptHandle interface {
+	// PodNominator abstracts operations to maintain nominated Pods.
+	PodNominator
+	// PluginsRunner abstracts operations to run some plugins.
+	PluginsRunner
+	// Extenders returns registered scheduler extenders.
+	Extenders() []SchedulerExtender
+}
+
+// PodNominator abstracts operations to maintain nominated Pods.
+type PodNominator interface {
+	// UpdateNominatedPodForNode adds the given pod to the nominated pod map or
+	// updates it if it already exists.
+	UpdateNominatedPodForNode(pod *v1.Pod, nodeName string)
+	// DeleteNominatedPodIfExists deletes nominatedPod from internal cache. It's a no-op if it doesn't exist.
+	DeleteNominatedPodIfExists(pod *v1.Pod)
+	// NominatedPodsForNode returns nominatedPos on the given node.
+	NominatedPodsForNode(nodeName string) []*v1.Pod
+}
+
+// PluginsRunner abstracts operations to run some plugins.
+// It's mainly used in PreFilter, Filter and PostFilter phases.
+type PluginsRunner interface {
```

Comment:
Can the `Framework` interface extend this interface?  

And can we plan to expose `Score` related runner thorough this interface??  As I mentioned [here](https://github.com/kubernetes/kubernetes/issues/90025#issuecomment-613781984), although I think no one has been convinced yet unfortunately, I would like `preemptHandle` to provide `Score` related runner.  By the way, `preemptHandle`  provides extenders that have `Prioritize` extenders.  So, I think we better to expose `Score` related runner though the interface for uniformity. 
 
If not,  how about renaming this with `FilterPluginsRunner` or similar?

WDYT??

Reply:
> preemptHandle also provides extenders that have Prioritize extenders.

🤔 I think the preemptHandle will just use the function `ProcessPreemption`.

> And can we plan to expose Score related runner thorough this interface?? As I mentioned here, although I think no one has been convinced yet unfortunately, I would like let preemptHandle provide Score related runners.

I am not sure if we need `Score` in `Preempt`. And it is worried that the preemption phase will take more time, resulting in the overall scheduling performance degradation. And also if we do that, how to deal with the strategies in `pickOneNodeForPreemption` and the result of `Score` plugins？ @everpeace

Reply:
>  I think the preemptHandle will just use the function ProcessPreemption.

Oh, really? I expected that `preemptHandle` will be exposed to the `PreemptionExtention` which I expect to be implemented in `PostFilter` plugin.  So, I supposed that preemption plugins implemented by users would be able to call extenders' methods freely.

> And it is worried that the preemption phase will take more time, resulting in the overall scheduling performance degradation. 

Even if we exposed `Score` related runners though this interface,  default `PreemptionExtention` won't call these runners in it, right?  So, I'm not sure why we should worry about performance degradation because of it.  I think we would need enough interface for implementers to implement a flexible preemption extension point.  I also think we should warn implementers that calling `Score` runners will affect scheduler performance, though.

Let me try to explain my motivation again (sorry for my poor explanation).  For example, assume some profile is configured with score functions below:

- `A` and `B` which comes from built-in
- `X` and `Y` which comes from implemented by some OSS 
- extender `E` having `Prioritize` function
- `S` and `T` which is implemented by self-developed plugins

And given preemptor pod `p` and if your preemption plugin found

- node `N1` can be schedulable upon preempting  victim `v1`, `v2` and `v3` 
  - let `P(N1)` be a set of pods _currently scheduled_ in `N1` (which includes victims)
- node `N2` can be schedulable upon preempting victim `w1` and `w2`  
  - let `P(N2)` be a set of pods _currently scheduled_ in `N2` (which includes victims)

Then, I feel that it would be natural that preemption plugin implementors want to compare

- normalized `NodeScore(N1)` with pods `P(N1) - {v1, v2, v3} + p` and 
- normalized `NodeScore(N2)` with with `P(N2) - {w1, w2} + p`

because the normalized scores by configured  score functions (`A, B, X, Y, E, S, and T` in this example) define the most preferred node to schedule preemptor pod `p`, right?

So, it would be so useful for preemption plugin implementors to be able to access`Score` runners that can easily calculate the normalized score by configured score functions.

> how to deal with the origin strategies in pickOneNodeForPreemption and the result of Score plugins?

sorry, I'm not sure what your point really is.  If extension implementers wanted to use the function, I don't disagree to expose this function through `preemptHandle`.  It would be not difficult because `pickOneNodeForPreemption` is now a stateless function.

Reply:
I can see @everpeace 's point. Adding `RunScorePlugins` into `preemptHanlde` can give out-of-tree plugins more flexibility to implement their preemption plugins, without reinventing existing Score algorithms. And default preemption plugin don't need to use it.

Reply:
@everpeace I updated the logic to use the private `f framework` to back `PluginsRunner` (instead of defining individual runXYZ methods), so that it'd be convenient for you to adopt RunFilterXYZ method by just introducing those methods to `PluginsRunner`.

Reply:
Thanks! LGTM!